hivemq-edge/src/main/java/com/hivemq/api/auth/provider/impl/ldap/LdapConnectionProperties.java

```diff
@@ -0,0 +1,349 @@
+        }
+
+        return options;
+    }
+
+    /**
+     * Creates a new LDAP connection using the configured properties.
+     * <p>
+     * The connection type depends on the configured TLS mode:
+     * <ul>
+     *     <li>NONE: Plain LDAP connection without encryption</li>
+     *     <li>LDAPS: TLS connection established from the start</li>
+     *     <li>START_TLS: Plain connection upgraded to TLS using StartTLS</li>
+     * </ul>
+     *
+     * @return A new LDAPConnection instance
+     * @throws LDAPException            if the connection fails
+     * @throws GeneralSecurityException if there's an SSL/TLS issue
+     */
+    public @NotNull LDAPConnection createConnection() throws LDAPException, GeneralSecurityException {
+        final LDAPConnectionOptions connectionOptions = createConnectionOptions();
+
+        return switch (tlsMode) {
+            case NONE -> createPlainConnection(connectionOptions);
+            case LDAPS -> createLdapsConnection(connectionOptions);
+            case START_TLS -> createStartTlsConnection(connectionOptions);
+        };
+    }
+
+    /**
+     * Creates a plain LDAP connection without encryption.
+     */
+    private @NotNull LDAPConnection createPlainConnection(final @NotNull LDAPConnectionOptions options)
+            throws LDAPException {
+        return new LDAPConnection(options, host, port);
+    }
+
+    /**
+     * Creates an LDAPS connection (TLS from start).
+     */
+    private @NotNull LDAPConnection createLdapsConnection(final @NotNull LDAPConnectionOptions options)
+            throws LDAPException, GeneralSecurityException {
+        final SSLContext sslContext = createSSLContext();
+        return new LDAPConnection(sslContext.getSocketFactory(), options, host, port);
+    }
+
+    /**
+     * Creates a connection and upgrades it to TLS using StartTLS.
+     */
+    private @NotNull LDAPConnection createStartTlsConnection(final @NotNull LDAPConnectionOptions options)
+            throws LDAPException, GeneralSecurityException {
+        // First create plain connection
+        final LDAPConnection connection = new LDAPConnection(options, host, port);
+
+        try {
+            // Upgrade to TLS using StartTLS extended operation
+            final SSLContext sslContext = createSSLContext();
+            final StartTLSExtendedRequest startTLSRequest = new StartTLSExtendedRequest(sslContext);
+            final ExtendedResult startTLSResult = connection.processExtendedOperation(startTLSRequest);
+
+            if (startTLSResult.getResultCode() != ResultCode.SUCCESS) {
+                throw new LDAPException(startTLSResult.getResultCode(),
+                        "StartTLS failed: " + startTLSResult.getDiagnosticMessage());
+            }
+
+            return connection;
+        } catch (final LDAPException | GeneralSecurityException e) {
+            // Close the connection if StartTLS fails
+            connection.close();
```

Comment:
Same question. Could it be moved to `finally`?

Reply:
finally doesn't really make sense since it would close the connection even on success.